lib/artifact.js

```diff
@@ -17,40 +17,80 @@ module.exports = (logger, requestOptions, requestPromise, responseHandler, Metad
 
 
         /**
-         * Uploads content to Shelf at the provided path.
+         * Uploads content to Shelf at the path this artifact was
+         * initialized with.
          *
-         * @param {(string|ReadStream)} content
-         * @return {Promise.<Object>}
+         * @param {string} content
+         * @return {Promise.<string>} location The URL the artifact was
+         *                                     uploaded to.
          */
         upload(content) {
-            var options;
+            var promise;
 
             logger.debug(`In ${this.constructor.name}.upload()`);
+            promise = bluebird.fromCallback((resolver) => {
+                var form, options, post;
+
+                options = requestOptions.createOptions(this.authToken);
+                options.url = this.uri;
+                logger.debug(`POST request options: ${JSON.stringify(options, null, 4)}`);
+                post = request.post(options, resolver);
+                form = post.form();
+                form.append("file", content, {
+                    filename: "file"
+                });
+            });
+
+            return promise.then((response) => {
+                return response.headers.location;
+            });
+        }
+
+
+        /**
+         * Uploads the given file to Shelf at the path this artifact was
+         * initialized with. If the parameter is a string, it will be treated
+         * as a local file system path, and will turn that path into a
+         * ReadStream. If the parameter is a ReadStream, it will be used
+         * directly.
+         *
+         * @param {(string|ReadStream)} file
+         * @return {Promise.<string>}
+         */
+        uploadFromFile(file) {
+            var options;
+
+            logger.debug(`In ${this.constructor.name}.uploadFromFile()`);
+
+            if (typeof file === "string") {
+                file = fs.createReadStream(file);
+            }
+
             options = requestOptions.createOptions(this.authToken);
             options.url = this.uri;
+            logger.debug(`POST request options: ${JSON.stringify(options, null, 4)}`);
             options.formData = {
-                file: {
-                    value: content,
-                    options: {
-                        filename: "file",
-                        contentType: "application/octet-stream"
-                    }
-                }
+                file
             };
-            logger.debug(`POST request options: ${JSON.stringify(options, null, 4)}`);
 
-            return requestPromise.post(options).then((response) => {
-                logger.debug(`In ${this.constructor.name}.upload() -> request.post().then()`);
+            return new Promise((resolve, reject) => {
```

Comment:
This doesn't seem to be a case where you couldn't just `return bluebird.<reject or resolve>` rather than wrapping the request in a `new Promise`.